### DISCUSSION TITLE
including a template into a Page

### DISCUSSION THREAD

**Discussion by mpounsett** [getpelican/pelican#3076] [Q&A] [answered]
I'm trying to do an RST include in a page, where I want to pull in a `jinja|html` file that is inside my theme.  The specific use case is loading my little social-media sidebar block into the Contact Me page.

``` rst
.. raw:: html
   :file: theme/mytheme/templates/include/social.html
```

The problem I'm having is that the parser can't find the HTML file.  Neither the documentation sections on linking to static files nor including content in RST or Markdown files mention what the CWD should be, or how (or if) you can include template files.  Can someone enlighten me?

Or is there a better approach to doing this?

**Comment by mpounsett** [comment]
Aha!  Probably should have continued with the trial-and-error for longer... found that it's in the `content/pages/` directory.

However, it looks like this include happens outside whatever part of Pelican does jinja template processing, because that part of the include file is just treated as raw text, regardless of whether I use `.. include:: [path]` or `.. raw:: html`

So hopefully someone can suggest a better approach for getting templated data into a Page.

**Comment by Epy** [comment]
Hello
I have a similar project for a Pelican website, with Piwigo for a page of images, not developped at all for now but I thought about it.

I was thinking to execute a python script gathering data and feeding it in a Jinja template, copied at the right place for Pelican to upload it with the html generated content.
If I'm able to go further, I'll make a Pelican plugin which will generate a full rst page, which will be used as the other content by Pelican. 

That what I would suggest: generating the entire page with your data in rst and copying it in `content` directory

**Comment by mpounsett** [comment, marked as answer]
After a bit more poking, I think I have found _The Way This is Supposed to Be Done_: `TEMPLATE_PAGES`.

The documentation as-written leads to some uncertainty about how to set this up though, and the default configuration leads to some misleading errors... so here's what I ended up doing to make this work:

1. Make sure that both `ARTICLE_PATHS` and `PAGE_PATHS` are set to search a directory under `PATH`, and are not empty variables.   

   If either are left blank then Pelican will find the not-RST-nor-MD files referenced by `TEMPLATE_PAGES` and there will be errors about missing document headers (like `title`).  By default, `PAGE_PATHS` is set but `ARTICLE_PATHS` is empty.  For example:  

   ```python
   ARTICLE_PATHS = ['articles']
   PAGE_PATHS = ['pages']
   ```

2. Create a third directory for documents under `PATH` (e.g. `content/template_pages/`).
3. Put your pure `jinja|html` document in that directory and reference it in `TEMPLATE_PAGES` with a source and destination.  The following example will pull a page from `content/template_pages/contact.html`, compile it, and then write it to `output/contact.html`:  

   ``` python
   TEMPLATE_PAGES = {
    'template_pages/contact.html': 'contact.html',
   }
   ```

   The CWD while compiling these seems to be the same as for any template in your theme.  This will be important if you're trying to include any pages from your theme, as I am.  Assuming you have `theme/mytheme/templates/includes/social.html` the above `contact.html` could contain:

   ```jinja
   {% include 'includes/social.html' %}
   ```
